### ISSUE TITLE
[JavaScript SDK Web][Adaptive Card Designer] [Designer does not render custom elements registered in adaptive cards]

### ISSUE THREAD

**Issue by Trish92** [microsoft/AdaptiveCards#5652]
# Web JS SDK

What platform is your issue or question related to? (Delete other platforms).

- [ ] JavaScript

# Author or host
All

# Version of SDK

What version are you using? Latest

# Details

Explain your issue. If you just have a question, please post [on StackOverflow instead](https://stackoverflow.com/questions/tagged/adaptive-cards).

https://docs.microsoft.com/en-us/adaptive-cards/sdk/rendering-cards/javascript/extensibility

This documentation shows an example to show a progress bar and create your own action
But how do I then show those in the designer ? Where can I find documentation for the same?

For the ProgressBarPeer, this is my class type : export class ProgressBarPeer extends ACDesigner.TypedCardElementPeer

This is what I sent to registerPeer : ACDesigner.CardDesignerSurface.cardElementPeerRegistry.registerPeer(
ProgressBar,
ProgressBarPeer,
"Elements",
'acd-icon-columnSet'
);
If I use AC.TextBlock instead of my own custom component ProgressBar I do see changes in the properties as defined in my ProgressBarPeer

Please let me know what maybe missing to see the ProgressBar under elements in the designer. I am using all the latest npm packages.
I am attaching as evidence my registry that contains an action and an element with their peers.
It would be great if there was some more documentation and an example for adding an action and an element to the designer. Thanks.

![image](https://user-images.githubusercontent.com/15216947/114421707-94899e00-9b83-11eb-932e-b27cdc993fa0.png)

#5240 
![image](https://user-images.githubusercontent.com/15216947/114421677-8b98cc80-9b83-11eb-904c-322c0ffd2722.png)

**Comment by Trish92** [comment]
https://stackoverflow.com/questions/66680403/adding-custom-card-elements-into-the-adaptive-card-designer

**Comment by jonmill** [comment]
@dclaux - can you please investigate and respond back with the right action we should take?

**Comment by Trish92** [comment]
You can reproduce this easily by using your own ProgressBar example and the TextBlockPeer file. I would appreciate any resolution you can provide cause this is very important at the moment for us. This project is amazing, we love the extensibility of it. Thank you so much!

**Comment by Trish92** [comment]
#5240

**Comment by dclaux** [comment]
I'm not sure how I can help you without seeing your code. Every single element in the designer is implemented exactly as you would if you were to create your own custom element. It requires an element registered with the renderer, and a designer peer registered with the designer. I am working on a custom element myself right with n associated designer peer and things work as expected.

We do lack proper documentation at the moment on how to extend the designer, we'll get to that. Please share you ProgressBarPeer implementation as well as the code you use to register both the ProgressBar element and the peer in order for me to be able to diagnose. Also make sure that the target version selected in the designer isn't lower than the version for which your element/peer are registered.

**Comment by Trish92** [comment]
`import * as Adaptive from 'adaptivecards';
import { ProgressBar } from './ProgressBar';
import * as ACDesigner from 'adaptivecards-designer';

export abstract class CardElementPeerInplaceEditor<
  TCardElement extends Adaptive.CardElement
> extends ACDesigner.DesignerPeerInplaceEditor {
  readonly cardElement: TCardElement;

  constructor(cardElement: TCardElement) {
    super();

    this.cardElement = cardElement;
  }
}

export class ProgressBarPeer extends ACDesigner.TypedCardElementPeer<ProgressBar> {
  static readonly textProperty = new ACDesigner.StringPropertyEditor(
    Adaptive.Versions.v1_3,
    'text',
    'Text',
    true,
    true
  );
  static readonly wrapProperty = new ACDesigner.BooleanPropertyEditor(Adaptive.Versions.v1_3, 'wrap', 'Wrap');
  static readonly maxLinesProperty = new ACDesigner.NumberPropertyEditor(
    Adaptive.Versions.v1_3,
    'maxLines',
    'Maximum lines',
    0
  );
  static readonly fontTypeProperty = new ACDesigner.EnumPropertyEditor(
    Adaptive.Versions.v1_3,
    'fontType',
    'Font type',
    Adaptive.FontType
  );
  static readonly sizeProperty = new ACDesigner.EnumPropertyEditor(
    Adaptive.Versions.v1_3,
    'size',
    'Size',
    Adaptive.TextSize
  );
  static readonly weightProperty = new ACDesigner.EnumPropertyEditor(
    Adaptive.Versions.v1_3,
    'weight',
    'Weight',
    Adaptive.TextWeight
  );
  static readonly colorProperty = new ACDesigner.EnumPropertyEditor(
    Adaptive.Versions.v1_3,
    'color',
    'Color',
    Adaptive.TextColor
  );
  static readonly subtleProperty = new ACDesigner.BooleanPropertyEditor(Adaptive.Versions.v1_3, 'isSubtle', 'Subtle');

  // protected createInplaceEditor(): ACDesigner.DesignerPeerInplaceEditor {
  //     return new ACDesigner.TextBlockPeerInplaceEditor(this.cardElement);
  // }

  protected internalGetTreeItemText(): string {
    return 'blue';
  }

  protected internalAddCommands(context: ACDesigner.DesignContext, commands: Array<ACDesigner.PeerCommand>) {
    super.internalAddCommands(context, commands);

    if (context.dataStructure) {
      commands.push(
        new ACDesigner.PeerCommand({
          name: 'Bind...',
          alwaysShowName: true,
          toolTip: 'my element.',
          execute: (command: ACDesigner.PeerCommand, clickedElement: HTMLElement) => {
            let fieldPicker = new ACDesigner.FieldPicker(context.dataStructure);
            fieldPicker.onClose = (sender, wasCancelled) => {
              if (!wasCancelled) {
                this.changed(true);
              }
            };
            fieldPicker.popup(clickedElement);
          },
        })
      );
    }
  }

  populatePropertySheet(
    propertySheet: ACDesigner.PropertySheet,
    defaultCategory: string = ACDesigner.PropertySheetCategory.DefaultCategory
  ) {
    super.populatePropertySheet(propertySheet, defaultCategory);

    propertySheet.add(defaultCategory, ProgressBarPeer.textProperty);

    propertySheet.add(
      ACDesigner.PropertySheetCategory.LayoutCategory,
      ProgressBarPeer.wrapProperty,
      ProgressBarPeer.maxLinesProperty
    );

    propertySheet.add(ACDesigner.PropertySheetCategory.StyleCategory, ProgressBarPeer.fontTypeProperty);
  }

  getToolTip(): string {
    return 'Double click to edit';
  }
  initializeCardElement() {
    super.initializeCardElement();
    if (!this.cardElement.title || this.cardElement.title == '') {
      this.cardElement.title = 'New TextBlock';
    }
  }
}
`

**Comment by Trish92** [comment]
[ProgressBarPeer.txt](https://github.com/microsoft/AdaptiveCards/files/6321478/ProgressBarPeer.txt)

**Comment by Trish92** [comment]
[ProgressBar.txt](https://github.com/microsoft/AdaptiveCards/files/6321480/ProgressBar.txt)

**Comment by Trish92** [comment]
If I use this same peer and register Adaptive.RichTextBlock or some other component I do see the property sheet as implemented in the ProgressBarPeer file

**Comment by Trish92** [comment]
`
 AC.GlobalRegistry.elements.register(ProgressBar.JsonTypeName, ProgressBar, AC.Versions.v1_3);
ACDesigner.CardDesignerSurface.cardElementPeerRegistry.registerPeer(ProgressBar,ProgressBarPeer,"Elemnts", "acd-icon-columnSet");`

**Comment by dclaux** [comment]
ProgressBar's properties are `title` and `value`. Your ProgressBarPeer declares `wrapProperty`, `fontTypeProperty`, `sizeProperty`, etc. These properties are irrelevant to ProgressBar. Your peer needs to declare properties for `title` and `value`.

**Comment by Trish92** [comment]
![image](https://user-images.githubusercontent.com/15216947/114954485-b0619e00-9e28-11eb-876b-814f5deb1013.png)
 I still dont see a progress bar element in the designer but if   I use Adaptive.RichTextBlock then I do see my added value and title properties

**Comment by Trish92** [comment]
![image](https://user-images.githubusercontent.com/15216947/114955311-71344c80-9e2a-11eb-90e2-862b6393ab15.png)
Difference between ProgressBar and another component

**Comment by dclaux** [comment]
Sorry I am confused as to what you're showing or what you're trying to explain.

To add support for a specific element in the designer, you need to implement a peer that declares the properties of the element you want to see in the property editor. For ProgressBar you need a ProgressBarPeer that declares title and value properties and you need to add those properties to the property sheet in `populatePropertySheet`. That's really about it.

**Comment by Trish92** [comment]
[ProgressBarPeer.txt](https://github.com/microsoft/AdaptiveCards/files/6321781/ProgressBarPeer.txt)

**Comment by Trish92** [comment]
I did do that but was unable to render the Progress bar element in the Elements section of the designer. But If I replace the ProgressBar that I made with Adaptive.RichTextBlock that Peer file does work and you can see a Title and Value property on the property sheet

**Comment by dclaux** [comment]
Are you registering the ProgressBar element in the `initialize` method of your `HostContainer`?

**Comment by Trish92** [comment]
nope that maybe the problem, I can try that . 

This is all I am doing 

 let hostContainers: any = ACDesigner.defaultMicrosoftHosts;

    // Optional: add the default Microsoft Host Apps (see docs below)
    hostContainers = [];

    this.designer = new ACDesigner.CardDesigner(hostContainers);

**Comment by Trish92** [comment]
Do you have an implementation handy?

**Comment by Trish92** [comment]
I am registering it at the start of the application before rendering the designer and adaptive card and I do see the component rendering in the adaptive card, just not in the designer.

**Comment by dclaux** [comment]
We really do need documentation, I'll try to find some time soon.

In the designer, which elements are supported (and thus displayed) is tied to the selected "host app". Under the covers, a host app is implemented by creating a class that extends the abstract HostContainer class. That base class declares its own element and action registries.

If you do not pass any host container to the constructor of the designer, it uses a default host container that only exposes built-in elements. If you need to use custom elements (and/or actions) in the designer you need to provide your own host container implementation, and the initialize method of that host container is where you can register your custom elements/actions.

It's very simple:
```typescript
class YourHostContainer extends DefaultContainer {
    initialize() {
        super.initialize();

        this.elementsRegistry.register(ProgressBar.JsonTypeName, ProgressBar);
    }
}

let designer = new CardDesigner([ new YourHostContainer ]);

```

**Comment by dclaux** [comment]
Or you should even be able to do this:
```typescript
let hostContainer = new DefaultContainer("your container name", "<path-to-its-css-stylesheet>");
hostContainer.elementsRegistry.register(ProgressBar.JsonTypeName, ProgressBar);

let designer = new CardDesigner([ hostContainer ]);
```

**Comment by Trish92** [comment]
Amazing ! Worked, thank you so very much ! 

`let hostContainer = new ACDesigner.DefaultContainer("MyHost", "../../styles/custom-styles.css");
    hostContainer.elementsRegistry.register(ProgressBar.JsonTypeName, ProgressBar);
    this.designer = new ACDesigner.CardDesigner([hostContainer]);`


You already have most of it documented :)

**Comment by Trish92** [comment]
I do have my custom element now rendering in the designer but unfortunately if I do update its properties, try to drag it etc. none of that works. It is not updating within the card payload editor. And even if I save the payload it doesnt contain my Progress Bar element. Is there something we need to do to enable the new component to get added to the adaptive card JSON and show up in the Card Payload Editor?

**Comment by Trish92** [comment]
I believe I got it, added this to progress bar

protected populateSchema(schema: AC.SerializableObjectSchema) {
    super.populateSchema(schema);
}

**Comment by dclaux** [comment]
That can't be it. This code is totally unnecessary since by default `super.populateSchema()` is called automatically.

**Comment by Trish92** [comment]
Ya, that is not it. My ProgressBar works fine but the new element I am creating is not working. I do see a typescript error in     
    
 hostContainer.elementsRegistry.register("name", Component)

**Comment by Trish92** [comment]
I am doing dependency injection in my custom component, at this time that is the only thing remaining to test.

**Comment by dclaux** [comment]
There's really no reason why it should be that hard. Would you mind sharing your code with me?

**Comment by Trish92** [comment]
I found the problem! I am using a node_module import in my project to get the custom class component because we have two projects : one with the designer embedded and the other is the one that actually renders the card. So to avoid duplication and future use cases in other projects I created a component. That works as expected up untill this point perhaps due to this method 

 ACDesigner.CardDesignerSurface.cardElementPeerRegistry.registerPeer(
      ProgressBar,
      ProgressBarPeer,
      'Elements',
      'acd-icon-columnSet'
    );
